### ISSUE TITLE
Fix build warnings after new connection API changes

### ISSUE THREAD

**Issue by arichiardi** [clojure-emacs/cider#2350]
Just a reminder for myself of whoever can do it first.

## Expected behavior
No warnings.

## Actual behavior
```
In end of data:
cider-connection.el:589:1:Warning: the function ‘cider-mode’ is not known to
    be defined.

...

In end of data:
cider-repl.el:1742:1:Warning: the following functions are not known to be
    defined: cider-jack-in-clojure, cider-jack-in-clojurescript,
    cider-connect-clojure, cider-connect-clojurescript,
    cider-connect-sibling-clojure, cider-connect-sibling-clojurescript
```
## Steps to reproduce the problem
make build

### CIDER version information

```
;; CIDER 0.18.0snapshot latest master
```

**Comment by bbatsov** [comment]
Yeah, that's a good point. Those are all from `cider.el` which we can't really include there. Normally we solve those by just using `declare-function`, but we should probably move the jack-in functionality to some "reusable" source file. Right now it certainly feels to me that the connection management/creation and REPL creation are a bit tangled. //cc @vspinu

**Comment by vspinu** [comment]
> we should probably move the jack-in functionality to some "reusable" source file.

That would basically mean renaming cider.el  :) Virtually all the variables and function in cider.el have something to do with the user level jack-in.

> connection management/creation and REPL creation 

I was troubled with this for a while till I realized that cider-repl.el is entirely about cider-repl-mode (or equivalently "user interaction with the repl"). `cider-connection.el` is all about lifecycle management of connections (which includes creation and deletion of connections buffers, aka repls). I have already added a note about this distinction into both files, but we could rename cider-repl.el into cider-repl-mode.el to emphasize this distinction even more.

**Comment by bbatsov** [comment]
> That would basically mean renaming cider.el :) Virtually all the variables and function in cider.el have something to do with the user level jack-in.

That file exists mostly as a single `require` point for the end users. The fact that some code is still in it is just an artefact of the old days - `nrepl.el` used to be a single-file project, which I first broke down to nrepl.el (for the repl stuff), nrepl-client.el and nrepl-interaction.el (for the nrepl-interactive-mode stuff). 

> I was troubled with this for a while till I realized that cider-repl.el is entirely about cider-repl-mode (or equivalently "user interaction with the repl"). cider-connection.el is all about lifecycle management of connections (which includes creation and deletion of connections buffers, aka repls). I have already added a note about this distinction into both files, but we could rename cider-repl.el into cider-repl-mode.el to emphasize this distinction even more.

Yeah, that was the intention. Although some of the things in `cider-repl.el` have to be shared elsewhere as well (apart from the REPL creation stuff, there's also everything that does output in a REPL buffer for instance). 

I'm much more worried about the source files having clear focus and no circular deps than about their names, though. Every `declare-function` that was added to avoid a circular dependency `require` is a `FIXME` reminder for me. The only valid use of `declare-function` is leverage auto-loading for certain rarely used functionality.
